spec/services/settings/language_update_service_spec.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+# -- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2023 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+# ++
+#
+
+require 'spec_helper'
+require_relative 'shared/shared_call_examples'
+require_relative 'shared/shared_setup_context'
+
+RSpec.describe Settings::LanguageUpdateService do
+  include_context 'with update service setup'
+
+  before do
+    # stub forcing default language
+    allow(instance)
+      .to receive(:force_user_language)
+  end
+
+  describe '#call' do
+    subject { instance.call(params) }
+
+    include_examples 'successful call'
+
+    it 'sets language of users having a non-available language to the default language' do
+      subject
+
+      expect(instance)
+        .to have_received(:force_user_language)
+    end
+
+    context 'when the contract is not successfully validated' do
+      let(:contract_success) { false }
+
+      include_examples 'unsuccessful call'
+    end
+  end
+end
```

Comment:
I won't mind if you let the test like it is, just please add the `expect(instance).not_to have_received(:force_user_language)` as it is really what is different between the two cases. Testing for successful/unsuccesful call is less important (and it's almost automatic as it is baked into the service code).

Reply:
> I felt a bit that since this is inheriting the entire behavior from Settings::UpdateService, that re-testing any of what is already being tested over on the specs for the base service is redundant and just makes the suite more fragile.

Agreed.

> I will be adding the test for not resetting the default language when the contract fails

Great :+1: 

> other than that, do you feel this would be clearer if I didn't abstract the identical setup code into a shared_context?

Yes probably. I just gave it a try, and I came down with this:

```ruby
require 'spec_helper'

RSpec.describe Settings::LanguageUpdateService do
  subject(:service) do
    described_class.new(user: build_stubbed(:admin))
  end

  before do
    allow(service).to receive(:force_user_language)
  end

  it 'sets language of users having a non-available language to the default language' do
    service.call('available_languages' => ['de', 'fr'])

    expect(service)
      .to have_received(:force_user_language)
  end

  context 'when the contract is not successfully validated' do
    before do
      allow(service).to receive(:validate_contract).and_return(ServiceResult.failure(message: 'fake error'))
    end

    it 'does not change language of any users' do
      service.call('available_languages' => ['de', 'fr'])

      expect(service)
        .not_to have_received(:force_user_language)
    end
  end
end
```

It feels more descriptive and simpler to understand in my opinion. And it uses less mocking which makes it a little slower.

Reply:
On the other hand, it's only testing that a method has been called. The test to check that the language of users is actually changed is in another file (`spec/controllers/admin/settings/languages_settings_controller_spec.rb`). I wonder if this test is useful then.

Reply:
I guess it might have some use in testing the alternate code path of not resetting users' languages unless successful and the controller test does only concern itself with the "happy path"?

Reply:
I think it's ok like this.
But generally, I like the unit test to test the thing properly, and then a higher-level test that tests the same code path could use some mocks, as it knows that the thing it mocks is already working as expected.

In this case, I think the test of `Settings::LanguageUpdateService` should test with real users that their language is reset, and the one testing `Admin::Settings::LanguagesSettingsController` simply ensuring that a `Settings::LanguageUpdateService` is instantiated and called is enough.

But in the end, both work.

Reply:
I'll keep this train of thought in mind; it makes sense in my opinion. Thanks for the having this discussion with me :)